File: src/main/java/com/untouchable/everytime/User/Controller/AuthController.java
package com.untouchable.everytime.User.Controller;

import com.untouchable.everytime.User.DTO.AuthResponse;
import com.untouchable.everytime.User.DTO.LoginRequest;
import com.untouchable.everytime.User.DTO.SignUpRequest;
import com.untouchable.everytime.User.Service.AuthService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;

@Tag(name = "1. Authentication API", description = "로그인, 회원 가입, 토큰 재발급 API (Authentication Header Should be Empty)")
@RestController
@RequiredArgsConstructor
@RequestMapping("/auth")
public class AuthController {

    /**
     * AuthService
     */
    private final AuthService authService;

    /**
     * 로그인 API
     *
     * @param loginRequest 로그인 요청 DTO
     * @return ResponseEntity<AuthResponse>
     */
    @PostMapping("/login")
    @Operation(summary = "로그인 API", description = "ID, Password 이용한 로그인 API<br>성공 시 AccessToken, RefreshToken 반환")
    public ResponseEntity<AuthResponse> login(
            @RequestBody LoginRequest loginRequest
    ) {
        return ResponseEntity.ok(authService.login(loginRequest));
    }


    /**
     * 회원가입 API
     *
     * @param signUpRequest 회원가입 요청 DTO
     * @return ResponseEntity<AuthResponse>
     */
    @PostMapping("/signup")
    @Operation(summary = "회원가입 API", description = "ID, Password, Name, Email 이용한 회원가입 API<br>성공 시 AccessToken, RefreshToken 반환")
    public ResponseEntity<AuthResponse> signup(
            @RequestBody SignUpRequest signUpRequest
    ) {
        return ResponseEntity.ok(authService.signup(signUpRequest));
    }

    /**
     * ID 중복체크 API
     *
     * @param userId 사용자 ID
     * @return ResponseEntity<Boolean>
     */
    @PostMapping("/checkId")
    @Operation(summary = "ID 중복체크 API", description = "사용자 ID 중복 체크 API<br>중복 시 True, 중복이 아닐 시 False 반환")
    public ResponseEntity<Boolean> checkId(
            @RequestParam(value = "userId") String userId
    ) {
        return ResponseEntity.ok(authService.userDuplicationCheck(userId));
    }


    /**
     * 토큰 재발급 API
     *
     * @param refreshToken Refresh Token
     * @return ResponseEntity<AuthResponse> (AccessToken Only)
     */
    @PostMapping("/refresh-token")
    @Operation(summary = "토큰 재발급 API", description = "Refresh Token을 이용한 AccessToken 재발급 API")
    public ResponseEntity<AuthResponse> refreshToken(
            @RequestHeader("RefreshToken") @Schema(description = "Refresh Token") String refreshToken
    ) {
        return ResponseEntity.ok(authService.refreshToken(refreshToken));
    }

    /**
     * 로그아웃 API
     *
     * @param principal Principal
     * @return ResponseEntity<String>
     */
    @PreAuthorize("isAuthenticated()")
    @PostMapping("/logout")
    @Operation(summary = "로그아웃 API", description = "로그아웃 API<br>저장된 Refresh Token들을 모두 삭제합니다 (모든 기기 로그아웃)")
    public ResponseEntity<String> logout(
            Principal principal
    ) {
        authService.logout(principal.getName());
        return ResponseEntity.ok("로그아웃 성공");
    }
}

File: src/main/java/com/untouchable/everytime/auth/controller/AuthController.kt
package com.untouchable.everytime.auth.controller

import com.untouchable.everytime.auth.controller.data.IdCheckForm
import com.untouchable.everytime.auth.controller.data.LoginForm
import com.untouchable.everytime.auth.controller.data.RegisterForm
import com.untouchable.everytime.auth.service.AuthService
import com.untouchable.everytime.auth.service.data.TokenResponse
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@Tag(name = "1. Auth", description = "인증 관련 API")
@RestController
@RequestMapping("/v1/api/auth")
class AuthController(
    private val authService: AuthService
) {

    @PostMapping("/id-check")
    @Operation(summary = "아이디 중복 확인", description = "True: 이미 사용 중, False: 사용 가능")
    fun checkId(
        @RequestBody form: IdCheckForm
    ): ResponseEntity<Boolean> {
        return ResponseEntity.ok(authService.checkId(form.id))
    }

    @PostMapping("/login")
    @Operation(summary = "로그인", description = "로그인 성공 시 토큰 반환")
    fun login(
        @RequestBody form: LoginForm
    ): ResponseEntity<TokenResponse> {
        return ResponseEntity.ok(authService.login(form))
    }

    @PostMapping("/register")
    @Operation(summary = "회원가입", description = "회원가입 성공 시 토큰 반환")
    fun register(
        @RequestBody form: RegisterForm
    ): ResponseEntity<TokenResponse> {
        return ResponseEntity.ok(authService.register(form))
    }
}
